clients/src/main/java/org/apache/kafka/common/requests/LeaderAndIsrRequest.java

```diff
@@ -121,8 +122,9 @@ public LeaderAndIsrRequest(Struct struct, short version) {
             List<Integer> replicas = new ArrayList<>(replicasArray.length);
             for (Object r : replicasArray)
                 replicas.add((Integer) r);
+            boolean isNew = partitionStateData.hasField(IS_NEW_KEY_NAME) ? partitionStateData.getBoolean(IS_NEW_KEY_NAME) : false;
```

Comment:
Discussed offline. I have updated the patch so that if there is IOException when reading brokerId from metadata file, the corresponding log dir will be marked failure but the broker will not shutdown. 

On the other hand, after broker registers itself in the zookeeper, it will only try to checkpoint brokerId in the metadata files in the live log directory (i.e. `logManager.liveLogDirs`). Most likely there won't be IOException when writing to metadata files here. If there is, then broker will shutdown entirely.